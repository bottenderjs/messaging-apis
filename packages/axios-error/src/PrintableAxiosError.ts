import util from 'util';

import { AxiosError } from 'axios';
import { JsonValue } from 'type-fest';

function indent(str: string): string {
  return str
    .split('\n')
    .map((s) => (s ? `  ${s}` : ''))
    .join('\n');
}

function json(data: JsonValue): string {
  return JSON.stringify(data, null, 2);
}

export default class PrintableAxiosError extends Error {
  /**
   * The config object of the error request.
   */
  public readonly config: AxiosError['config'];

  /**
   * The request object of the error request.
   */
  public readonly request?: AxiosError['request'];

  /**
   * The response object of the error request.
   */
  public readonly response?: AxiosError['response'];

  /**
   * The response status of the error request.
   */
  public readonly status?: number;

  /**
   * @example
   * ```js
   * new AxiosError(errorThrownByAxios)
   * ```
   */
  constructor(error: AxiosError);

  /**
   * @example
   * ```js
   * new AxiosError('error message', errorThrownByAxios)
   * ```
   */
  constructor(message: string, error: AxiosError);

  /**
   * @example
   * ```js
   * new AxiosError('error message', { config, request, response })
   * ```
   */
  constructor(
    message: string,
    error: Pick<AxiosError, 'config' | 'request' | 'response'>
  );

  constructor(
    messageOrError: string | AxiosError,
    error?: AxiosError | Pick<AxiosError, 'config' | 'request' | 'response'>
  ) {
    let err: Pick<AxiosError, 'config' | 'request' | 'response'>;
    let message: string;
    if (typeof messageOrError === 'string') {
      message = messageOrError;
      err = error as Pick<AxiosError, 'config' | 'request' | 'response'>;
    } else {
      message = messageOrError.message;
      err = messageOrError;
    }

    super(message);

    const { config, request, response } = err;

    this.config = config;
    this.request = request;
    this.response = response;
    if (response && response.status) {
      this.status = response.status;
    }

    this.name = 'PrintableAxiosError';
  }

  [util.inspect.custom](): string {
    let requestMessage = '';

    if (this.config) {
      let { data } = this.config;

      try {
        data = JSON.parse(data);
      } catch (_) {} // eslint-disable-line

      let requestData = '';

      if (this.config.data) {
        requestData = `
Request Data -
${indent(json(data))}`;
      }

      requestMessage = `
Request -
  ${this.config.method ? this.config.method.toUpperCase() : ''} ${
        this.config.url
      }
${requestData}`;
    }

    let responseMessage = '';

    if (this.response) {
      let responseData;

      if (this.response.data) {
        responseData = `
Response Data -
${indent(json(this.response.data))}`;
      }

      responseMessage = `
Response -
  ${this.response.status} ${this.response.statusText}
${responseData}`;
    }

    return `
${this.stack}

Error Message -
  ${this.message}
${requestMessage}
${responseMessage}
`;
  }
}
