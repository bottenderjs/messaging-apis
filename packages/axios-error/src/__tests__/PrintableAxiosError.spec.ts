import util from 'util';

import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';

import PrintableAxiosError from '../PrintableAxiosError';

const mock = new MockAdapter(axios);

mock.onAny().reply(400, {
  error_status: 'boom....',
});

const stack = `Error: boom....
  at Object.<anonymous> (/Users/xxx/messaging-apis/packages/axios-error/src/__tests__/index.spec.js:16:19)
  at Generator.throw (<anonymous>)
  at step (/Users/xxx/messaging-apis/packages/axios-error/src/__tests__/index.spec.js:4:336)
  at /Users/xxx/messaging-apis/packages/axios-error/src/__tests__/index.spec.js:4:535
  at <anonymous>
`;

it('should work', async () => {
  try {
    await axios.post('/', { x: 1 });
  } catch (err) {
    // overwrite because axios-mock-adapter set it to undefined
    err.response.statusText = 'Bad Request';

    const error = new PrintableAxiosError(err.response.data.error_status, err);

    // overwrite stack to test it
    error.stack = stack;

    expect(error[util.inspect.custom]()).toMatchSnapshot();
    expect(error.name).toBe('PrintableAxiosError');
  }
});

it('should set `.status` property', async () => {
  try {
    await axios.post('/', { x: 1 });
  } catch (err) {
    // overwrite because axios-mock-adapter set it to undefined
    err.response.statusText = 'Bad Request';

    const error = new PrintableAxiosError(err);

    expect(error.status).toBe(400);
  }
});

it('should work with construct using error instance only', async () => {
  try {
    await axios.post('/', { x: 1 });
  } catch (err) {
    // overwrite because axios-mock-adapter set it to undefined
    err.response.statusText = 'Bad Request';

    const error = new PrintableAxiosError(err);

    // overwrite stack to test it
    error.stack = stack;

    expect(error[util.inspect.custom]()).toMatchSnapshot();
    expect(error.name).toBe('PrintableAxiosError');
  }
});

it('should work with undefined response', async () => {
  try {
    await axios.post('/', { x: 1 });
  } catch (err) {
    // overwrite to undefined
    // https://github.com/Yoctol/bottender/issues/246
    err.response = undefined;

    const error = new PrintableAxiosError('read ECONNRESET', err);

    // overwrite stack to test it
    error.stack = stack;

    expect(error[util.inspect.custom]()).toMatchSnapshot();
    expect(error.name).toBe('PrintableAxiosError');
  }
});
