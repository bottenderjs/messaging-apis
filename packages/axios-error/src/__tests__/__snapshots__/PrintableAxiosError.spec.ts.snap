// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should work 1`] = `
"
Error: boom....
  at Object.<anonymous> (/Users/xxx/messaging-apis/packages/axios-error/src/__tests__/index.spec.js:16:19)
  at Generator.throw (<anonymous>)
  at step (/Users/xxx/messaging-apis/packages/axios-error/src/__tests__/index.spec.js:4:336)
  at /Users/xxx/messaging-apis/packages/axios-error/src/__tests__/index.spec.js:4:535
  at <anonymous>


Error Message -
  boom....

Request -
  POST /

Request Data -
  {
    \\"x\\": 1
  }

Response -
  400 Bad Request

Response Data -
  {
    \\"error_status\\": \\"boom....\\"
  }
"
`;

exports[`should work with construct using error instance only 1`] = `
"
Error: boom....
  at Object.<anonymous> (/Users/xxx/messaging-apis/packages/axios-error/src/__tests__/index.spec.js:16:19)
  at Generator.throw (<anonymous>)
  at step (/Users/xxx/messaging-apis/packages/axios-error/src/__tests__/index.spec.js:4:336)
  at /Users/xxx/messaging-apis/packages/axios-error/src/__tests__/index.spec.js:4:535
  at <anonymous>


Error Message -
  Request failed with status code 400

Request -
  POST /

Request Data -
  {
    \\"x\\": 1
  }

Response -
  400 Bad Request

Response Data -
  {
    \\"error_status\\": \\"boom....\\"
  }
"
`;

exports[`should work with undefined response 1`] = `
"
Error: boom....
  at Object.<anonymous> (/Users/xxx/messaging-apis/packages/axios-error/src/__tests__/index.spec.js:16:19)
  at Generator.throw (<anonymous>)
  at step (/Users/xxx/messaging-apis/packages/axios-error/src/__tests__/index.spec.js:4:336)
  at /Users/xxx/messaging-apis/packages/axios-error/src/__tests__/index.spec.js:4:535
  at <anonymous>


Error Message -
  read ECONNRESET

Request -
  POST /

Request Data -
  {
    \\"x\\": 1
  }

"
`;
